### ISSUE TITLE
is it possible to get card data to a local json file?

### ISSUE THREAD

**Issue by seiyria** [ronelm2000/wsmtools#52]
hi,

I'm building a [suite of tools](https://limelight.cards/) and I'm looking to find a good way to automate getting EN card data. I've tried building my own scraper, but it looks like there's one here that works. I just can't seem to figure out the documentation on how I could use it for things that aren't deck exports.

my goal: call `./wstools --card-id=5HY/W83-E001 --json=cache/5HY_W83_E001.json` and get a json file with all of the relevant card information

(I already plan to handle other aspects of the pipeline)

let me know if this is possible and I just missed it!

**Comment by ronelm2000** [comment]
Not exactly, there is a database parser, but a JSON-based deck parser is not yet supported (only one for Cockatrice). One other issue is that `IDatabaseExportInfo` currently only supports `ReleaseID`, which is usually `W##` (where ## is the number of the set of the JP/EN set and W pertains to the side and language -- Bushiroad has done several formats for this which can make it really confusing depending on the type of set), so those will have to be further supported down the line.

I'm actually currently trying to add in compiling YYT pricing data tho, so this will be on the works.

**Comment by ronelm2000** [comment]
PS: these are all covered under `./wstools export-db --rids <rid1;rid2> --exporter <exporter-alias>`, which hasn't really seen much support; I plan on changing that in the future as the deck export function has been really stable which is frees up devtime for things like exporting to CSV and unifying APIs for further card game projects.

**Comment by seiyria** [comment]
I see! I'll be eagerly awaiting progress in this regard then. It would be great to not have to, as you say, deal with Bushis confusing setup. Thanks for the quick response!

**Comment by ronelm2000** [comment]
I forgot that I also need to add the proper documentation for this functionality (it's missing on releases.)

Considering this only had exactly one use case, I couldn't be bothered to add it before this.

**Comment by ronelm2000** [comment]
You can test the JSON Exporter format in this nightly build and see if it suits your needs.

```ps
# Parses Bang Dream 5th Anniv from HOTC, then exports it to result.json
./wstools parse "https://heartofthecards.com/translations/bang_dream_girls_band_party!_5th_anniversary_booster_pack.html" --with noskip:yyt
./wstools export-db --rids W95 --exporter json
```

https://github.com/ronelm2000/wsmtools/actions/runs/3443476892

**Comment by seiyria** [comment]
Is there no way to use a single set id and get it from bushi or HOTC? Right now, what I have is [sets and card ids per set](https://github.com/LimeLightCards/scraper-data), and if possible, I would like to retrieve them individually. Ideally, I wouldn't want to use any URLs.

Thanks for your work on this so far.

**Comment by ronelm2000** [comment]
> Is there no way to use a single set id and get it from bushi or HOTC? 
From both of them individually, no. I think the only one that can do that in a consistent way is via the EncoreDecks API and DeckLog API, which I have indeed integrated in the Deck-related commands, but not necessarily on its own commands.

You also tend to lose alot of information using Bushiroad only; for example using Bushiroad's DeckLog API only gives you the name, rarity, serial, and image link, with no information about its traits, effects, etc.

You can get sets via EncoreDecks API (which is the recommended way and is used automatically if you export a deck), but not all the sets are there (especially if they're new, really old, or straight up PRs with serial that are out there, in which case you tend to not have the English translation.)

There is no way to get all set URLs from HOTC apart from straight up web scraping the entire site, which is not recommended as a use case for automation.

If you find something else (like a new image repo with a corresponding API), you can make another Issue so that I can integrate it, but for now, information related to data on Weiss Schwarz publicly available in APIs is sparse.

**Comment by seiyria** [comment]
Hrm. I was having ok success using this: https://en.ws-tcg.com/cardlist/list/?cardno=MM/W17-E001 (I couldn't get decklog to work, actually), but HOTC also supports individual card links: https://www.heartofthecards.com/code/cardlist.html?card=WS_AB/W11-101

My main hangup with the first link I shared is that the data came out inconsistently at times, and I never spent a ton of time on HOTC.

I guess what I'm not sure about is if you're parsing sets, why looking up individual cards is a stretch to add? 

I realize above I misspoke; I meant _card id_ not _set id_. I can do without anything related to sets; I already scrape wstcg.com to get all of the card ids.

**Comment by ronelm2000** [comment]
It's because each page takes alot of resource (because of resolving HTML and manually adding cookies which is how DeckLog images can be fetched properly), and minimizing the amount of memory used for each command. Not to mention HOTC has a hard limit on the amount of pages you can fetch in 10 seconds; it is absolutely not recommended to fetch each card link manually.

For example, to fetch a certain card promo like [5HY/W83-P01](https://heartofthecards.com/code/cardlist.html?card=WS_5HY/W83-P01), with just the serial, the first thing that needs to be done is to ensure that the card exists without fetching the data off HOTC, this means fetching data off DeckLog first, but do that with 10 URLs and you're gonna hit the limit extremely fast. This is why the recommended way is to use EncoreDecks which is connected to a separate manually maintained database and has publicly available API, which presents a slight issue: it doesn't have any way to get individual cards.

As for WS TCG URLs like  https://en.ws-tcg.com/cardlist/list/?cardno=MM/W17-E001 ; this was added as a way to fetch English-specific promo cards; since HOTC already has set translation pages, there was no need to support individual cards in the first place.

Anyway, all these should be resolvable with #54

**Comment by ronelm2000** [comment]
Oh right, I almost forgot that I should add support for `--serials` for the next update.

**Comment by seiyria** [comment]
Hmmmm. I see. I suppose I can wait for #54. In fairness I'm okay with rate limits - I can build that in. But, I realize my specific use case is not quite what most people would be going for. 

I actually was running into CORS issues with encoredecks, trying to build an integration to link to "decks that use <x> card" so I wasn't sure what it all supported. I can reach out again and maybe he would be okay with adding some of that [edit: I reached out to him in discord, he said he would add an endpoint for getting card by id]. I'd guess in a console setting instead of web setting though CORS won't be a problem.

Either way, thank you for all the explanation! I can definitely wait - and based on what I saw in #54 I can fetch multiple cards by id too if I separate them with semicolons, which is something I can work with. But, my ultimate end goal is at least the ability to look up singular cards in case I end up having holes in my data.

If it makes sense that all of this would be resolved by #54 then this can be closed.

Cheers!
